Cython/Compiler/ExprNodes.py

```diff
@@ -12213,6 +12213,35 @@ def c_types_okay(self, type1, type2):
             and (type2.is_int or type2.is_enum)
 
 
+class BitwiseOrNode(IntBinopNode):
+    #  '|' operator.
+
+    def analyse_pytyping_modifiers(self, env):
+        if self.operand1.is_none or self.operand2.is_none:
+            return ['typing.Optional']
+
+    def _analyse_bitwise_or_none(self, env, operand_node, leftop_is_none):
+        """Analyse annotations in form `[...] | None` and `None | [...]`"""
+        ttype = operand_node.analyse_as_type(env)
+        if not ttype:
+            return
+        if not ttype.can_be_optional():
+            # If ttype cannot be optional we need to return equivalent type allowing None values. If
+            # such type does not exist we must error out.
+            if ttype.equivalent_type and not operand_node.as_cython_attribute():
+                return ttype.equivalent_type
+            else:
+                signature = "None | [...]" if leftop_is_none else "[...] | None"
+                error(operand_node.pos, f"{signature} cannot be applied to type {ttype}")
```

Comment:
I don't think we need to be specific about the operator order in the error message. Both orders are semantically the same. Usually, I'd expect users to write `xyz | None`, everyone else will probably understand the message either way. We can then also avoid the `leftop_is_none` flag.
```suggestion
                error(operand_node.pos, f"'[...] | None' cannot be applied to type {ttype}")
```